Membrane components, mechanism, and examples

zjuradoq opened the pull request:
This pull request contains membrane components, mechanisms, and a specialized example for each type. Additionally, the component was updated to have a compartment attribute. Unit tests for each new component and mechanism have also been included. This pull request supersedes previous pull requests, which aimed to add membrane components.

This request also is in place of the 1/16/2025 pull request.

ayush9pandey reviewed (commented):
The notebook runs fine (and all examples are great!) I have some code-specific comments that I have added to the files directly. Some refactoring may be needed there. I also noticed some unit tests are failing: test_membrane_signaling and test_simple_diffusion (variables not defined). 

I can take a look at the PR again and possibly also try to refactor some code soon, if needed.

zjuradoq commented on biocrnpyler/components_membrane.py:
```diff
@@ -0,0 +1,356 @@
+    def __init__(self, membrane_protein: Union[Species, str, Component],
+                 product: Union[Species,str, Component],
+                 direction= None, size=None, attributes=None, **keywords):       
+        """Initialize a MembraneChannel object to store membrane channel related information.
```
Yes, it should be "Initialize an IntegralMembraneProtein component" and has been updated.

zjuradoq replied:
"Direction" is a string, and will be in species attribute. Is this a clarifying question or do I need to address an issue?

zjuradoq commented on biocrnpyler/components_membrane.py:
```diff
@@ -0,0 +1,356 @@
+
+    def update_reactions(self):
+        mech_cat = self.get_mechanism('membrane_insertion')
+        print(self)
```
This print and others have been removed. They were part of the debugging process.

zjuradoq commented on biocrnpyler/components_membrane.py:
```diff
@@ -0,0 +1,356 @@
+                 substrate: Union[Species, str, Component],
+                 direction=None, internal_compartment='Internal', external_compartment='External',
+                 ATP=None, cell=None, attributes=None, **keywords):
+        """Initialize a Transporter object to store Transport membrane related information.
```
Yes, I used a general term as a transport but changed it to a specific object, as Transporter is not a class.

zjuradoq commented on biocrnpyler/components_membrane.py:
```diff
@@ -0,0 +1,356 @@
+                 substrate: Union[Species, str, Component],
+                 direction=None, internal_compartment='Internal', external_compartment='External',
+                 cell=None, attributes=None, **keywords):
+        """Initialize a Transporter object to store transport membrane related information.
```
I have updated the language to the specific class.

zjuradoq commented on biocrnpyler/mechanisms_signaling.py:
```diff
@@ -0,0 +1,222 @@
+        Mechanism.__init__(self, name, mechanism_type)
+    
+    def update_species(self, membrane_sensor_protein, response_protein, 
+                       assigned_substrate, signal_substrate, energy, waste,
```
Would adding a dict i.e. 
`Complexes={"Signal_Sub:Sensor_Protein":complex1, "Signal_Sub:Sensor_Protein:ATP":complex2,...` be sufficient? Then, would this dictionary be used when calling the complexes in the reactions?

zjuradoq commented on biocrnpyler/mechanisms_transport.py:
```diff
@@ -0,0 +1,365 @@
+        if part_id is None and component is not None:
+            part_id = component.name
+    
+        if component is None and (kb1 is None or ku1 is None or kb2 is None or ku2 is None or kcat is None or kex is None):
```
It appears to be a remnant from a previous line; ku2 is not a parameter in the membrane protein integration reactions.

zjuradoq commented on biocrnpyler/mechanisms_transport.py:
```diff
@@ -0,0 +1,365 @@
+            
+        return [integral_membrane_protein,  product, complex1, complex2]
+    
+    def update_reactions(self, integral_membrane_protein, product, component=None, part_id=None, complex=None, complex2 = None, 
```
If you have time to meet, I would appreciate it, not sure how best to resolve this comment.

zjuradoq commented:
Updated the branch membrane components, mechanisms, and examples. All checks have run successfully!
The major update addresses the comment of creating dictionaries for reactions and complexes for better readability of the mechanisms with multiple complexes. Due to the changes in defining the complexes, examples and unit tests also had to be updated.

ayush9pandey reviewed (approved):
All changes look good! Thank you @zjuradoq!

ayush9pandey commented:
Merging...
